---
title: "Analyze user input dataset"
author: "Jie Tan"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    fig_width: 7
    fig_height: 7
vignette: >
  %\VignetteIndexEntry{User input example}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

This is an example for analyzing a user-provided microarray dataset.

# Data preparation
Load in required libraries.

```{r}
library("ADAGEpath")
library("dplyr")
library("DT")
library("knitr")
```

Before any analysis, we need to specify the ADAGE model and the data compendium
we want to use.

```{r}
model <- eADAGEmodel
compendium <- PAcompendium
probe_dist <- probedistribution
```

Let's load in a sample dataset that come with the package. It's expression data
from *Pseudomonas aeruginosa* wild type and ∆anr grown as biofilms on ∆F508 cystic
fibrosis bronchial epithelial cells (CFBEs). Detailed information about
the dataset can be found here
[GSE67006](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE67006).
All its CEL files are stored in the folder "./inst/extdata/anr".
To load your own dataset, simply modify this input path.

```{r}
input_path <- system.file("extdata", "anr/", package = "ADAGEpath")
data_raw <- load_dataset(input = input_path, isProcessed = FALSE,
                         isRNAseq = FALSE, model = model,
                         compendium = compendium, quantile_ref = probe_dist,
                         norm01 = FALSE)
```

ADAGE only accepts expression values in the (0,1) range. We linearly transform
expression values to be between 0 and 1 using the Pa compendium as the reference.

```{r}
data_normed <- zeroone_norm(input_data = data_raw, use_ref = TRUE,
                            ref_data = compendium)
```

Now let's specify the phenotypes for each sample. It needs to be a character
vector and has the same sample order as the expression data loaded above.

```{r}
data_pheno <- c("mt", "mt", "mt", "wt", "wt", "wt")
```

# ADAGE signature analysis

## Activity calculation
We calculate the activity of each signature for each sample in the dataset.

```{r}
data_activity <- calculate_activity(input_data = data_normed, model = model)
```

The returned `data_activity` is a `data.frame` with signature names in the first
column and activity values per sample starting from the second column.

## Active signature detection
We want to find signatures that are differentially active between
anr mutant and wildtype samples.
We use [limma](https://bioconductor.org/packages/release/bioc/html/limma.html)
to perform a differential activation test. limma is more robust
than a simple t test when sample size is small. A two-group limma analysis is
provided in the function `build_limma()`. You can also build other limma models
to test signatures' activities when the experimental design is more complex.

In the limma test, we will use "wt" as the control phenotype.
Because there are a lot of signatures passing the significance cutoff, here we
use the more strigent bonferroni procedure instead of Benjamini–Hochberg
procedure for multiple hypothesis correction.

```{r}
limma_result <- build_limma(data_activity, phenotypes = data_pheno,
                            control_pheno = "wt",
                            use.bonferroni = TRUE)
```

To take both absolute activity difference and significance into account, we
use pareto fronts to pick the most differentially active signatures. We extract
differentially active signatures in the first 10 layers of pareto
fronts. Modify N_fronts to get more or fewer signatures.

```{r}
active_sigs <- get_active_signatures(limma_result = limma_result,
                                     pheno_group = "both", 
                                     method = "pareto", N_fronts = 10)
```

Signatures that are differentially active between anr mutant and wildtype are:

```{r}
active_sigs
```

Check out each signature's activity changes and significance in the `limma` test.

```{r}
plot_volcano(limma_result, highlight_signatures = active_sigs,
             interactive = TRUE)
```

Look at how the activities of active signature vary across samples.

```{r}
plot_activity_heatmap(activity = data_activity, signatures = active_sigs)
```

Combining the volcano plot and the activity heatmap, Node35pos is the most
active signature in anr mutant, followed by Node233pos, Node140pos.
On the other side, Node38neg, Node31pos, Node269pos, Node205neg are most active
in wildtype.

## Signature overlap
To reduce the number of signatures to look at, we can check whether these active
signatures overlap with each other.

`plot_signature_overlap` creates a heatmap of odds ratios. The odds ratio
represents the odds that two signatures share a specific number of genes.

```{r}
signature_similarity <- plot_signature_overlap(selected_signatures = active_sigs,
                                               model = model)
```

Next we calculate the marginal activities of similar signatures. Marginal
activity is defined as the activity of signature A after removing genes that
share with signature B.

```{r}
marginal_activity <- calculate_marginal_activity(input_data = data_normed,
                                                 selected_signatures = active_sigs,
                                                 model = model)
```

Again, we build a limma model to test whether these marginal activities
are still strongly different between two conditions.

```{r}
marginal_limma <- build_limma(input_data = marginal_activity,
                              phenotypes = data_pheno, control_pheno = "wt")
```

Let's visualize the marginal activities in a matrix heatmap.  The value in this
matrix represents the -log10 transformed adjusted p value in
the activation test when the effect of the column signature is removed from
the row signature. Values in the diagonal of the heatmap are the activation
significance of signatures themselves.

```{r}
plot_marginal_activation(marginal_limma_result = marginal_limma,
                         signature_order = colnames(signature_similarity),
                         sig_cutoff = 0.05)

```

Based on this plot, we can see that Node119pos, Node214neg, and Node299pos are
completely masked by Node191pos, because after removing Node191pos from them,
their significance drops dramatically. Node130pos and Node250neg are masked by
Node31pos. Node285neg is masked by Node205neg.
Node154pos is masked by Node57neg. Node63neg, Node39neg, Node228pos, Node158neg,
Node140pos, Node269neg and Node31neg are masked by Node35pos. Node185neg and
Node275pos are masked by Node67pos. Node278pos is masked by Node9pos.
We can see that Node67pos, Node35pos, and Node233pos each has unique
genes that make them still significant even after removing the effect
of another signature. We can safely remove a signature being masked by another
signature as long as we keep the second signature.

```{r}
unique_active_sigs <- remove_redundant_signatures(marginal_limma,
                                                  sig_cutoff = 0.05)
unique_active_sigs
```

Check out each signature's activity changes and significance after removing
overlapping signatures.

```{r}
plot_volcano(limma_result, highlight_signatures = unique_active_sigs,
             interactive = TRUE)
```

Look at how the activities of active signature vary across samples after removing
overlapping signatures.

```{r}
plot_activity_heatmap(activity = data_activity, signatures = unique_active_sigs)
```

## Active signature annotation
We download *Pseudomonas aeruginosa* KEGG pathway terms from the
[TRIBE](http://tribe.greenelab.com/#/home) web server.

```{r}
KEGG <- fetch_geneset(type = "KEGG")
# we only consider KEGG pathways with more than 5 genes and less than 100 genes
# as meaningful pathways
KEGG_subset <- KEGG[lengths(KEGG) >= 5 & lengths(KEGG) <= 100]
```

We associate active signatures to known KEGG pathways.

```{r}
pathway_result <- annotate_signatures_with_genesets(
  selected_signatures = unique_active_sigs, model = model,
  genesets = KEGG_subset)
```

Look at the activity of associated pathways inside active signatures. This help
differentiate active vs. non-active pathways assoicated with active signatures.

```{r}
pathway_activity <- signature_geneset_activity(
  signature_geneset_df = pathway_result[, c("signature", "geneset")],
  gene_set_list = KEGG_subset, model = model, input_data = data_normed)
plot_activity_heatmap(pathway_activity)
```

We can run a limma test on pathway activities and only keep pathways that truly
active in this dataset.

```{r}
pathway_limma_result <- build_limma(pathway_activity, phenotypes = data_pheno,
                                    control_pheno = "wt", use.bonferroni = TRUE)
pathway_limma_result$signature <- sapply(rownames(pathway_limma_result),
                                         function(x) unlist(strsplit(x, "\\|"))[[1]])
pathway_limma_result$geneset <- sapply(rownames(pathway_limma_result),
                                       function(x) unlist(strsplit(x, "\\|"))[[2]])

# combine limma result with pathway result and only keep truly active pathways
pathway_result_sig <- dplyr::left_join(
  pathway_limma_result %>% select(signature, geneset, logFC, adj.P.Val) %>%
    filter(adj.P.Val <= 0.05),
  pathway_result %>% select(-pvalue),
  by = c("signature" = "signature", "geneset" = "geneset"))
pathway_result_sig <- pathway_result_sig %>% rename(
  `activity difference` = logFC, `gene set enrichment (adj.P.val)` = qvalue)
pathway_result_sig <- pathway_result_sig[order(pathway_result_sig$adj.P.Val), ]
knitr::kable(pathway_result_sig, digits = 4, row.names = FALSE, align = "c")
```

The following pathways are associated with active signatures but are not
active in this dataset themselves.

```{r}
setdiff(pathway_result$geneset, pathway_result_sig$geneset)
```

We also performed a GSEA analysis on this dataset and found that GSEA and
ADAGEpath agree on many pathways. Here we list pathways enriched in GSEA
but not associated with ADAGE signatures in the top 10 layers of pareto fronts.
We check whether they are assoicated with
any ADAGE signatures and how differentially active the signatures are if
associated.

```{r}
# pathways enriched in GSEA but not associated with ADAGE signatures
# in the top 10 layers of pareto fronts
GSEA_only <- c( "KEGG-Module-M00178", "KEGG-Pathway-pae03010",
                "KEGG-Pathway-pae03070", "KEGG-Module-M00331",
                "KEGG-Module-M00332", "KEGG-Module-M00023",
                "KEGG-Pathway-pae00280", "KEGG-Pathway-pae00970")
# get the names of all signatures
all_sigs <- c(paste0(colnames(model)[-1], "pos"),
              paste0(colnames(model)[-1], "neg"))
# peform pathway association for all signatures
all_enrichment <- annotate_signatures_with_genesets(selected_signatures = all_sigs,
                                                    model = model, genesets = KEGG_subset)
# for each pathway, get its associated signatures and record the one that is most
# significantly differentially active
associated_sigs <- sapply(GSEA_only, function(x){
  sigs <- all_enrichment$signature[grepl(x, all_enrichment$geneset)]
  if (identical(sigs, character(0))) {
    NA
  } else {
    sigs[which.min(limma_result[sigs, "adj.P.Val"])]
  }
})
# see how these signatures perform in the limma test
plot_volcano(limma_result, highlight_signatures = associated_sigs,
             interactive = TRUE)
```

These pathways enriched in the GSEA are not associated with any signatures:

```{r}
GSEA_only[is.na(associated_sigs)]
```

There are also signatures uncharacterized by KEGG.
```{r}
uncharacterized_sigs <- setdiff(unique_active_sigs, pathway_result$signature)
uncharacterized_sigs
```

Check the signature similarity of the uncharacterized signatures.
```{r}
plot_signature_overlap(selected_signatures = uncharacterized_sigs)
```

They form two major groups. Group1 contains Node38pos, Node67pos, Node35pos, and
Node233pos. Group2 contains Node191pos, Node269pos, and Node107pos.

## Gene-gene network
We can check how genes in the active signatures cluster in the ADAGE gene-gene
network.

We can calculate a expression fold change for each gene and pass it to the
gene-gene network to show as node color. Again, we use limma to test
differential expression and get the logFC.

```{r}
data_raw_limma <- build_limma(input_data = data_raw, phenotypes = data_pheno,
                              control_pheno = "wt")
# build a gene:fold change table from limma result
gene_logFC <- data.frame(geneID = rownames(data_raw_limma),
                         logFC = data_raw_limma$logFC)
```

Visualize the ADAGE gene-gene network of the active signatures.

```{r}
visualize_gene_network(selected_signatures = unique_active_sigs,
                       gene_color_value = gene_logFC,
                       model = model, cor_cutoff = 0.5,
                       curated_pathways = KEGG)
```

To review a signature, we can list all genes in it.

```{r}
DT::datatable(annotate_genes_in_signatures(selected_signatures = "Node155neg",
                                           model = model,
                                           curated_pathways = KEGG))
```

And visualize its genes with the gene-gene network.

```{r}
visualize_gene_network(selected_signatures = "Node155neg",
                       gene_color_value = gene_logFC,
                       model = model, cor_cutoff = 0.5,
                       curated_pathways = KEGG)
```

We can also review signatures in group.

```{r}
sig_group <- c("Node38pos", "Node67pos", "Node35pos", "Node233pos")
DT::datatable(annotate_genes_in_signatures(selected_signatures = sig_group,
                                           model = model,
                                           curated_pathways = KEGG))
```

```{r}
visualize_gene_network(selected_signatures = sig_group,
                       gene_color_value = gene_logFC,
                       model = model, cor_cutoff = 0.5,
                       curated_pathways = KEGG)
```
